File: HighFive/HighFive/System/IO/filestream.cs
// ==++==
//
//   Copyright (c) Microsoft Corporation.  All rights reserved.
//
// ==--==
/*============================================================
**
** Class:  FileStream
**
** <OWNER>Microsoft</OWNER>
**
**
** Purpose: Exposes a Stream around a file, with full
** synchronous and asychronous support, and buffering.
**
**
===========================================================*/
/*
 * https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/filestream.cs
 */

using System.Runtime.InteropServices;
using System.Threading.Tasks;

/*
 * FileStream supports different modes of accessing the disk - async mode
 * and sync mode.  They are two completely different codepaths in the
 * sync & async methods (ie, Read/Write vs. BeginRead/BeginWrite).  File
 * handles in NT can be opened in only sync or overlapped (async) mode,
 * and we have to deal with this pain.  Stream has implementations of
 * the sync methods in terms of the async ones, so we'll
 * call through to our base class to get those methods when necessary.
 *
 * Also buffering is added into FileStream as well. Folded in the
 * code from BufferedStream, so all the comments about it being mostly
 * aggressive (and the possible perf improvement) apply to FileStream as
 * well.  Also added some buffering to the async code paths.
 *
 * Class Invariants:
 * The class has one buffer, shared for reading & writing.  It can only be
 * used for one or the other at any point in time - not both.  The following
 * should be true:
 *   0 <= _readPos <= _readLen < _bufferSize
 *   0 <= _writePos < _bufferSize
 *   _readPos == _readLen && _readPos > 0 implies the read buffer is valid,
 *     but we're at the end of the buffer.
 *   _readPos == _readLen == 0 means the read buffer contains garbage.
 *   Either _writePos can be greater than 0, or _readLen & _readPos can be
 *     greater than zero, but neither can be greater than zero at the same time.
 *
 */

namespace System.IO
{
    [H5.External]
    [H5.Namespace(false)]
    internal class FileReader
    {
        public extern FileReader();

        [H5.Convention(H5.Notation.CamelCase)]
        public extern void ReadAsArrayBuffer(object file);

        [H5.Convention(H5.Notation.CamelCase)]
        public readonly string Result;

        [H5.Convention(H5.Notation.LowerCase)]
        public Action OnLoad;

        [H5.Convention(H5.Notation.LowerCase)]
        public Action<object> OnError;
    }

    [H5.Reflectable]
    [H5.Convention]
    public class FileStream : Stream
    {
        private string name;
        byte[] _buffer;

        public FileStream(string path, FileMode mode)
        {
            this.name = path;
        }

        internal FileStream(byte[] buffer, string name)
        {
            this._buffer = buffer;
            this.name = name;
        }

        internal static Task<FileStream> FromFile(object file)
        {
            var completer = new System.Threading.Tasks.TaskCompletionSource<FileStream>();
            var fileReader = new FileReader();
            /*@
            fileReader.onload = function () {
                completer.setResult(new System.IO.FileStream.ctor(fileReader.result, file.name));
            };
            */
            fileReader.OnError = (e) =>
            {
                completer.SetException(new SystemException(e.As<dynamic>().target.error.As<string>()));
            };
            fileReader.ReadAsArrayBuffer(file);

            return completer.Task;
        }

        public override bool CanRead
        {
            get
            {
                return true;
            }
        }

        public override bool CanWrite
        {
            get
            {
                return false;
            }
        }

        public override bool CanSeek
        {
            get
            {
                return false;
            }
        }

        public virtual bool IsAsync
        {
            get
            {
                return false;
            }
        }

        public String Name
        {
            get
            {
                return name;
            }
        }

        public override long Length
        {
            get
            {
                return this.GetInternalBuffer().ToDynamic().byteLength;
            }
        }

        public override long Position
        {
            get;
            set;
        }

        public override void Flush()
        {
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            throw new NotImplementedException();
        }

        public override void SetLength(long value)
        {
            throw new NotImplementedException();
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            throw new NotImplementedException();
        }

        private byte[] GetInternalBuffer()
        {
            if (this._buffer == null)
            {
                this._buffer = FileStream.ReadBytes(this.name);

            }

            return this._buffer;
        }

        internal async Task EnsureBufferAsync()
        {
            if (this._buffer == null)
            {
                this._buffer = await FileStream.ReadBytesAsync(this.name);
            }
        }

        public override int Read([In, Out] byte[] buffer, int offset, int count)
        {
            if (buffer == null)
            {
                throw new System.ArgumentNullException("buffer");
            }

            if (offset < 0)
            {
                throw new System.ArgumentOutOfRangeException("offset");
            }

            if (count < 0)
            {
                throw new System.ArgumentOutOfRangeException("count");
            }

            if ((buffer.Length - offset) < count)
            {
                throw new System.ArgumentException();
            }

            var num = this.Length - this.Position;
            if (num > count)
            {
                num = count;
            }

            if (num <= 0)
            {
                return 0;
            }

            var byteBuffer = H5.Script.Write<dynamic>("new Uint8Array(this.GetInternalBuffer())");
            if (num > 8)
            {
                for (var n = 0; n < num; n++)
                {
                    buffer[n + offset] = H5.Script.Write<dynamic>("byteBuffer[this.Position.add(System.Int64(n))]");
                }
            }
            else
            {
                var num1 = num;
                while (true)
                {
                    var num2 = num1 - 1;
                    num1 = num2;
                    if (num2 < 0)
                    {
                        break;
                    }
                    buffer[offset + num1] = H5.Script.Write<dynamic>("byteBuffer[this.Position.add(num1)]");
                }
            }
            this.Position += num;
            return (int)num;
        }

        internal static byte[] ReadBytes(string path)
        {
            if (H5.Script.IsNode)
            {
                var fs = H5.Script.Write<dynamic>(@"require(""fs"")");
                return H5.Script.Write<dynamic>("H5.cast(fs.readFileSync(path), ArrayBuffer)");
            }
            else
            {
                var req = H5.Script.Write<dynamic>("new XMLHttpRequest()");
                req.open("GET", path, false);
                req.overrideMimeType("text/plain; charset=x-user-defined");
                req.send(null);
                if (H5.Script.Write<bool>("req.status !== 200"))
                {
                    throw new IOException("Status of request to " + path + " returned status: " + req.status);
                }

                string text = req.responseText;
                var resultArray = H5.Script.Write<dynamic>("new Uint8Array(text.length)");
                text.ToCharArray().ForEach((v, index, array) => resultArray[index] = (byte)(v & byte.MaxValue));
                return resultArray.buffer;
            }
        }

        internal static Task<byte[]> ReadBytesAsync(string path)
        {
            var tcs = new TaskCompletionSource<byte[]>();

            if (H5.Script.IsNode)
            {
                var fs = H5.Script.Write<dynamic>(@"require(""fs"")");
                fs.readFile(path, new Action<object, byte[]>((err, data) => {
                    if (err != null)
                    {
                        throw new IOException();
                    }

                    tcs.SetResult(data);
                }));
            }
            else
            {
                var req = H5.Script.Write<dynamic>("new XMLHttpRequest()");
                req.open("GET", path, true);
                req.overrideMimeType("text/plain; charset=binary-data");
                req.send(null);

                /*@
                req.onreadystatechange = function () {
                */
                    if (H5.Script.Write<bool>("req.readyState !== 4"))
                    {
                        H5.Script.Write("return;");
                    }

                    if (H5.Script.Write<bool>("req.status !== 200"))
                    {
                        throw new IOException("Status of request to " + path + " returned status: " + req.status);
                    }

                    string text = req.responseText;
                    var resultArray = H5.Script.Write<dynamic>("new Uint8Array(text.length)");
                    text.ToCharArray().ForEach((v, index, array) => resultArray[index] = (byte)(v & byte.MaxValue));
                    tcs.SetResult(resultArray.buffer);
                /*@
                };
                */
            }

            return tcs.Task;
        }
    }
}

File: HighFive/HighFive/Resources/.generated/System/IO/FileStream.js
    HighFive.define("System.IO.FileStream", {
        inherits: [System.IO.Stream],
        statics: {
            methods: {
                FromFile: function (file) {
                    var completer = new System.Threading.Tasks.TaskCompletionSource();
                    var fileReader = new FileReader();
                    fileReader.onload = function () {
                        completer.setResult(new System.IO.FileStream.ctor(fileReader.result, file.name));
                    };
                    fileReader.onerror = function (e) {
                        completer.setException(new System.SystemException.$ctor1(HighFive.unbox(e).target.error.As()));
                    };
                    fileReader.readAsArrayBuffer(file);

                    return completer.task;
                },
                ReadBytes: function (path) {
                    if (HighFive.isNode) {
                        var fs = require("fs");
                        return HighFive.cast(fs.readFileSync(path), ArrayBuffer);
                    } else {
                        var req = new XMLHttpRequest();
                        req.open("GET", path, false);
                        req.overrideMimeType("text/plain; charset=x-user-defined");
                        req.send(null);
                        if (req.status !== 200) {
                            throw new System.IO.IOException.$ctor1(System.String.concat("Status of request to " + (path || "") + " returned status: ", req.status));
                        }

                        var text = req.responseText;
                        var resultArray = new Uint8Array(text.length);
                        System.String.toCharArray(text, 0, text.length).forEach(function (v, index, array) {
                                var $t;
                                return ($t = (v & 255) & 255, resultArray[index] = $t, $t);
                            });
                        return resultArray.buffer;
                    }
                },
                ReadBytesAsync: function (path) {
                    var tcs = new System.Threading.Tasks.TaskCompletionSource();

                    if (HighFive.isNode) {
                        var fs = require("fs");
                        fs.readFile(path, HighFive.fn.$build([function (err, data) {
                            if (err != null) {
                                throw new System.IO.IOException.ctor();
                            }

                            tcs.setResult(data);
                        }]));
                    } else {
                        var req = new XMLHttpRequest();
                        req.open("GET", path, true);
                        req.overrideMimeType("text/plain; charset=binary-data");
                        req.send(null);

                        req.onreadystatechange = function () {
                        if (req.readyState !== 4) {
                            return;
                        }

                        if (req.status !== 200) {
                            throw new System.IO.IOException.$ctor1(System.String.concat("Status of request to " + (path || "") + " returned status: ", req.status));
                        }

                        var text = req.responseText;
                        var resultArray = new Uint8Array(text.length);
                        System.String.toCharArray(text, 0, text.length).forEach(function (v, index, array) {
                                var $t;
                                return ($t = (v & 255) & 255, resultArray[index] = $t, $t);
                            });
                        tcs.setResult(resultArray.buffer);
                        };
                    }

                    return tcs.task;
                }
            }
        },
        fields: {
            name: null,
            _buffer: null
        },
        props: {
            CanRead: {
                get: function () {
                    return true;
                }
            },
            CanWrite: {
                get: function () {
                    return false;
                }
            },
            CanSeek: {
                get: function () {
                    return false;
                }
            },
            IsAsync: {
                get: function () {
                    return false;
                }
            },
            Name: {
                get: function () {
                    return this.name;
                }
            },
            Length: {
                get: function () {
                    return System.Int64(this.GetInternalBuffer().byteLength);
                }
            },
            Position: System.Int64(0)
        },
        ctors: {
            $ctor1: function (path, mode) {
                this.$initialize();
                System.IO.Stream.ctor.call(this);
                this.name = path;
            },
            ctor: function (buffer, name) {
                this.$initialize();
                System.IO.Stream.ctor.call(this);
                this._buffer = buffer;
                this.name = name;
            }
        },
        methods: {
            Flush: function () { },
            Seek: function (offset, origin) {
                throw new System.NotImplementedException.ctor();
            },
            SetLength: function (value) {
                throw new System.NotImplementedException.ctor();
            },
            Write: function (buffer, offset, count) {
                throw new System.NotImplementedException.ctor();
            },
            GetInternalBuffer: function () {
                if (this._buffer == null) {
                    this._buffer = System.IO.FileStream.ReadBytes(this.name);

                }

                return this._buffer;
            },
            EnsureBufferAsync: function () {
                var $step = 0,
                    $task1, 
                    $taskResult1, 
                    $jumpFromFinally, 
                    $tcs = new System.Threading.Tasks.TaskCompletionSource(), 
                    $returnValue, 
                    $async_e, 
                    $asyncBody = HighFive.fn.bind(this, function () {
                        try {
                            for (;;) {
                                $step = System.Array.min([0,1,2,3], $step);
                                switch ($step) {
                                    case 0: {
                                        if (this._buffer == null) {
                                            $step = 1;
                                            continue;
                                        } 
                                        $step = 3;
                                        continue;
                                    }
                                    case 1: {
                                        $task1 = System.IO.FileStream.ReadBytesAsync(this.name);
                                        $step = 2;
                                        if ($task1.isCompleted()) {
                                            continue;
                                        }
                                        $task1.continue($asyncBody);
                                        return;
                                    }
                                    case 2: {
                                        $taskResult1 = $task1.getAwaitedResult();
                                        this._buffer = $taskResult1;
                                        $step = 3;
                                        continue;
                                    }
                                    case 3: {
                                        $tcs.setResult(null);
                                        return;
                                    }
                                    default: {
                                        $tcs.setResult(null);
                                        return;
                                    }
                                }
                            }
                        } catch($async_e1) {
                            $async_e = System.Exception.create($async_e1);
                            $tcs.setException($async_e);
                        }
                    }, arguments);

                $asyncBody();
                return $tcs.task;
            },
            Read: function (buffer, offset, count) {
                if (buffer == null) {
                    throw new System.ArgumentNullException.$ctor1("buffer");
                }

                if (offset < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("offset");
                }

                if (count < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("count");
                }

                if ((((buffer.length - offset) | 0)) < count) {
                    throw new System.ArgumentException.ctor();
                }

                var num = this.Length.sub(this.Position);
                if (num.gt(System.Int64(count))) {
                    num = System.Int64(count);
                }

                if (num.lte(System.Int64(0))) {
                    return 0;
                }

                var byteBuffer = new Uint8Array(this.GetInternalBuffer());
                if (num.gt(System.Int64(8))) {
                    for (var n = 0; System.Int64(n).lt(num); n = (n + 1) | 0) {
                        buffer[System.Array.index(((n + offset) | 0), buffer)] = byteBuffer[this.Position.add(System.Int64(n))];
                    }
                } else {
                    var num1 = num;
                    while (true) {
                        var num2 = num1.sub(System.Int64(1));
                        num1 = num2;
                        if (num2.lt(System.Int64(0))) {
                            break;
                        }
                        buffer[System.Array.index(System.Int64.toNumber(System.Int64(offset).add(num1)), buffer)] = byteBuffer[this.Position.add(num1)];
                    }
                }
                this.Position = this.Position.add(num);
                return System.Int64.clip32(num);
            }
        }
    });
